src/librustc/ty/structural_impls.rs

```diff
@@ -850,28 +846,38 @@ impl<'tcx> TypeFoldable<'tcx> for ty::ProjectionPredicate<'tcx> {
 
 impl<'tcx> TypeFoldable<'tcx> for ty::ExistentialProjection<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
-        ty::ExistentialProjection {
-            trait_ref: self.trait_ref.fold_with(folder),
-            item_name: self.item_name,
-            ty: self.ty.fold_with(folder),
-        }
+        // FIXME(tschottdorf): simplify.
+        let folded_ty = self.ty.fold_with(folder);
+        let trait_ref = self.trait_ref(folder.tcx());
+        let item_name = self.item_name(folder.tcx());
+        let folded_trait_ref = trait_ref.fold_with(folder);
+        ty::ExistentialProjection::from_ty_ref_and_name(
+            folder.tcx(),
+            folded_ty,
+            folded_trait_ref,
+            item_name,
+        )
```

Comment:
This should be easy to fix because `trait_ref` here was `ExistentialTraitRef` so `visit_trait_ref` never got called.